File: apps/android/app/src/main/java/org/llamenos/hotline/ui/reports/ReportsViewModel.kt
package org.llamenos.hotline.ui.reports

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.filterNotNull
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.serialization.builtins.MapSerializer
import kotlinx.serialization.builtins.serializer
import kotlinx.serialization.json.Json
import org.llamenos.hotline.api.ApiService
import org.llamenos.hotline.api.SessionState
import org.llamenos.hotline.crypto.CryptoService
import org.llamenos.hotline.hub.ActiveHubState
import org.llamenos.hotline.model.AssignReportRequest
import org.llamenos.hotline.model.CreateReportRequest
import org.llamenos.hotline.model.CreateTypedReportRequest
import org.llamenos.hotline.model.Report
import org.llamenos.hotline.model.ReportCategoriesResponse
import org.llamenos.hotline.model.ReportEnvelope
import org.llamenos.hotline.model.CmsReportTypesResponse
import org.llamenos.hotline.model.ReportsListResponse
import org.llamenos.hotline.model.UpdateReportRequest
import org.llamenos.protocol.ReportTypeDefinition
import javax.inject.Inject

data class ReportsUiState(
    val reports: List<Report> = emptyList(),
    val total: Int = 0,
    val isLoading: Boolean = false,
    val isRefreshing: Boolean = false,
    val error: String? = null,
    val selectedStatus: ReportStatusFilter = ReportStatusFilter.ALL,
    val selectedCategory: String? = null,
    val categories: List<String> = emptyList(),
    val selectedReport: Report? = null,

    // Report types
    val reportTypes: List<ReportTypeDefinition> = emptyList(),
    val isLoadingReportTypes: Boolean = false,
    val reportTypesError: String? = null,
    val selectedReportType: ReportTypeDefinition? = null,
    val fieldValues: Map<String, String> = emptyMap(),

    // Action states
    val isCreating: Boolean = false,
    val isClaiming: Boolean = false,
    val isClosing: Boolean = false,
    val createSuccess: Boolean = false,
    val actionError: String? = null,
) {
    /**
     * Report types filtered to those marked as mobile-optimized.
     * Falls back to all non-archived types if none are mobile-optimized.
     */
    val mobileReportTypes: List<ReportTypeDefinition>
        get() {
            val activeTypes = reportTypes.filter { !it.isArchived }
            val mobileTypes = activeTypes.filter { it.mobileOptimized }
            return mobileTypes.ifEmpty { activeTypes }
        }

    /** Whether CMS report types are available for this hub. */
    val hasReportTypes: Boolean
        get() = reportTypes.any { !it.isArchived }
}

enum class ReportStatusFilter(val queryParam: String?) {
    ALL(null),
    ACTIVE("active"),
    WAITING("waiting"),
    CLOSED("closed"),
}

/**
 * ViewModel for the reports screen.
 *
 * Loads reports from GET /reports with status and category filtering.
 * Supports creating (both legacy and typed), claiming, and closing reports.
 * Fetches CMS report type definitions to drive template-based report creation.
 */
@HiltViewModel
class ReportsViewModel @Inject constructor(
    private val apiService: ApiService,
    private val cryptoService: CryptoService,
    private val sessionState: SessionState,
    private val activeHubState: ActiveHubState,
) : ViewModel() {

    private val _uiState = MutableStateFlow(ReportsUiState())
    val uiState: StateFlow<ReportsUiState> = _uiState.asStateFlow()

    private val json = Json {
        ignoreUnknownKeys = true
        encodeDefaults = true
    }

    init {
        activeHubState.activeHubId
            .filterNotNull()
            .onEach { loadReports() }
            .launchIn(viewModelScope)
    }

    fun loadReports() {
        viewModelScope.launch {
            _uiState.update {
                it.copy(
                    isLoading = it.reports.isEmpty(),
                    isRefreshing = it.reports.isNotEmpty(),
                    error = null,
                )
            }
            try {
                val query = buildString {
                    append(apiService.hp("/api/reports"))
                    append("?limit=50")
                    val status = _uiState.value.selectedStatus
                    if (status.queryParam != null) {
                        append("&status=${status.queryParam}")
                    }
                    val category = _uiState.value.selectedCategory
                    if (category != null) {
                        append("&category=$category")
                    }
                }
                val response = apiService.request<ReportsListResponse>("GET", query)
                _uiState.update {
                    it.copy(
                        reports = response.conversations,
                        total = response.total,
                        isLoading = false,
                        isRefreshing = false,
                    )
                }
            } catch (e: Exception) {
                _uiState.update {
                    it.copy(
                        isLoading = false,
                        isRefreshing = false,
                        error = e.message ?: "Failed to load reports",
                    )
                }
            }
        }
    }

    private fun loadCategories() {
        viewModelScope.launch {
            try {
                val response = apiService.request<ReportCategoriesResponse>("GET", apiService.hp("/api/reports/categories"))
                _uiState.update { it.copy(categories = response.categories) }
            } catch (_: Exception) {
                // Non-critical — categories are optional
            }
        }
    }

    /**
     * Fetch CMS report type definitions from the backend.
     *
     * Report types drive the template-based report creation flow.
     * If the hub has no CMS report types configured, the legacy
     * free-form report creation screen is used as a fallback.
     */
    fun loadReportTypes() {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoadingReportTypes = true, reportTypesError = null) }
            try {
                val response = apiService.request<CmsReportTypesResponse>(
                    "GET",
                    "/api/settings/cms/report-types",
                )
                _uiState.update {
                    it.copy(
                        reportTypes = response.reportTypes,
                        isLoadingReportTypes = false,
                    )
                }
            } catch (_: Exception) {
                // Non-critical — fall back to legacy report creation
                _uiState.update {
                    it.copy(
                        isLoadingReportTypes = false,
                    )
                }
            }
        }
    }

    /**
     * Select a report type for the typed report creation form.
     */
    fun selectReportType(reportType: ReportTypeDefinition) {
        _uiState.update {
            it.copy(
                selectedReportType = reportType,
                fieldValues = emptyMap(),
                actionError = null,
            )
        }
    }

    /**
     * Update a field value in the current typed report form.
     */
    fun updateFieldValue(fieldName: String, value: String) {
        _uiState.update {
            it.copy(fieldValues = it.fieldValues + (fieldName to value))
        }
    }

    /**
     * Clear the selected report type and field values.
     */
    fun clearReportType() {
        _uiState.update {
            it.copy(
                selectedReportType = null,
                fieldValues = emptyMap(),
            )
        }
    }

    fun refresh() {
        loadReports()
    }

    fun setStatusFilter(filter: ReportStatusFilter) {
        _uiState.update { it.copy(selectedStatus = filter) }
        loadReports()
    }

    fun setCategoryFilter(category: String?) {
        _uiState.update { it.copy(selectedCategory = category) }
        loadReports()
    }

    fun selectReport(report: Report) {
        _uiState.update { it.copy(selectedReport = report) }
    }

    /**
     * Create a new encrypted report (legacy flow without report type).
     *
     * Encrypts the report body using the same envelope pattern as notes,
     * then sends the title, optional category, and encrypted content to the API.
     */
    fun createReport(title: String, category: String?, body: String) {
        viewModelScope.launch {
            _uiState.update { it.copy(isCreating = true, actionError = null, createSuccess = false) }
            try {
                val encrypted = cryptoService.encryptNote(body, sessionState.adminPubkeys)
                val envelopes = encrypted.envelopes.map { env ->
                    ReportEnvelope(
                        pubkey = env.recipientPubkey,
                        wrappedKey = env.wrappedKey,
                        ephemeralPubkey = env.ephemeralPubkey,
                    )
                }
                val request = CreateReportRequest(
                    title = title,
                    category = category?.takeIf { it.isNotBlank() },
                    encryptedContent = encrypted.ciphertext,
                    readerEnvelopes = envelopes,
                )
                apiService.request<Report>("POST", apiService.hp("/api/reports"), request)
                _uiState.update { it.copy(isCreating = false, createSuccess = true) }
                refresh()
            } catch (e: Exception) {
                _uiState.update {
                    it.copy(
                        isCreating = false,
                        actionError = e.message ?: "Failed to create report",
                    )
                }
            }
        }
    }

    /**
     * Create a typed report using the selected report type template.
     *
     * Field values are serialized to JSON, encrypted with the same E2EE
     * envelope pattern, and submitted with the reportTypeId attached.
     *
     * @param reportTypeId The CMS report type ID
     * @param title The user-provided report title
     * @param fieldValues Map of field name to value (strings, including serialized selections)
     */
    fun createTypedReport(
        reportTypeId: String,
        title: String,
        fieldValues: Map<String, String>,
    ) {
        viewModelScope.launch {
            _uiState.update { it.copy(isCreating = true, actionError = null, createSuccess = false) }
            try {
                // Serialize field values as JSON for E2EE encryption
                val fieldsJson = json.encodeToString(
                    MapSerializer(String.serializer(), String.serializer()),
                    fieldValues,
                )

                val encrypted = cryptoService.encryptNote(fieldsJson, sessionState.adminPubkeys)
                val envelopes = encrypted.envelopes.map { env ->
                    ReportEnvelope(
                        pubkey = env.recipientPubkey,
                        wrappedKey = env.wrappedKey,
                        ephemeralPubkey = env.ephemeralPubkey,
                    )
                }

                val reportType = _uiState.value.reportTypes.find { it.id == reportTypeId }
                val request = CreateTypedReportRequest(
                    title = title,
                    category = reportType?.category?.value?.takeIf { it.isNotBlank() && it != "report" },
                    reportTypeId = reportTypeId,
                    encryptedContent = encrypted.ciphertext,
                    readerEnvelopes = envelopes,
                )
                apiService.request<Report>("POST", apiService.hp("/api/reports"), request)
                _uiState.update {
                    it.copy(
                        isCreating = false,
                        createSuccess = true,
                        fieldValues = emptyMap(),
                        selectedReportType = null,
                    )
                }
                refresh()
            } catch (e: Exception) {
                _uiState.update {
                    it.copy(
                        isCreating = false,
                        actionError = e.message ?: "Failed to create report",
                    )
                }
            }
        }
    }

    /**
     * Clear the create success flag after the UI has navigated back.
     */
    fun clearCreateSuccess() {
        _uiState.update { it.copy(createSuccess = false) }
    }

    /**
     * Claim a report by assigning it to the current volunteer.
     *
     * Sends POST /reports/:id/assign with the current user's pubkey,
     * which also transitions the report status from "waiting" to "active".
     */
    fun claimReport(reportId: String) {
        viewModelScope.launch {
            _uiState.update { it.copy(isClaiming = true, actionError = null) }
            val pubkey = cryptoService.pubkey
            if (pubkey == null) {
                _uiState.update {
                    it.copy(isClaiming = false, actionError = "No identity available")
                }
                return@launch
            }
            try {
                val request = AssignReportRequest(assignedTo = pubkey)
                val updated = apiService.request<Report>("POST", apiService.hp("/api/reports/$reportId/assign"), request)
                _uiState.update {
                    it.copy(
                        isClaiming = false,
                        selectedReport = updated,
                    )
                }
                refresh()
            } catch (e: Exception) {
                _uiState.update {
                    it.copy(
                        isClaiming = false,
                        actionError = e.message ?: "Failed to claim report",
                    )
                }
            }
        }
    }

    /**
     * Close a report by updating its status to "closed".
     *
     * Sends PATCH /reports/:id with { status: "closed" }.
     */
    fun closeReport(reportId: String) {
        viewModelScope.launch {
            _uiState.update { it.copy(isClosing = true, actionError = null) }
            try {
                val request = UpdateReportRequest(status = "closed")
                val updated = apiService.request<Report>("PATCH", apiService.hp("/api/reports/$reportId"), request)
                _uiState.update {
                    it.copy(
                        isClosing = false,
                        selectedReport = updated,
                    )
                }
                refresh()
            } catch (e: Exception) {
                _uiState.update {
                    it.copy(
                        isClosing = false,
                        actionError = e.message ?: "Failed to close report",
                    )
                }
            }
        }
    }

    /**
     * Dismiss the action error message.
     */
    fun dismissActionError() {
        _uiState.update { it.copy(actionError = null) }
    }

    fun dismissError() {
        _uiState.update { it.copy(error = null) }
    }
}

File: apps/ios/Sources/ViewModels/ReportsViewModel.swift
import Foundation
import UIKit

// MARK: - ReportsViewModel

/// View model for the Reports feature. Fetches reports from the API,
/// manages status filtering, and handles report creation with E2EE encryption.
/// Supports both legacy category-based reports and template-driven typed reports.
@Observable
final class ReportsViewModel {
    private let apiService: APIService
    private let cryptoService: CryptoService
    private let adminPubkeys: [String]

    // MARK: - Public State

    /// Reports from the server.
    var reports: [ClientReportResponse] = []

    /// Available report categories from the server.
    var categories: [String] = []

    /// Report type definitions fetched from CMS settings.
    /// Populated from `GET /api/settings/cms/report-types` (preferred) with fallback
    /// to `GET /api/reports/types` (legacy).
    var reportTypes: [ClientReportTypeDefinition] = []

    /// CMS report types fetched directly from the settings endpoint.
    /// Includes full CMS-specific fields (hubId, isSystem, numberingEnabled, etc.).
    var cmsReportTypes: [ClientReportTypeDefinition] = []

    /// Current status filter.
    var selectedFilter: ReportStatusFilter = .all

    /// Selected report type filter (nil = all types).
    var selectedTypeFilter: String?

    /// Whether the initial load is in progress.
    var isLoading: Bool = false

    /// Whether a report action (create, claim, close) is in progress.
    var isActionInProgress: Bool = false

    /// Error message from the last failed operation.
    var errorMessage: String?

    /// Whether the report creation sheet is shown.
    var showCreateSheet: Bool = false

    /// Whether the report type picker is shown.
    var showReportTypePicker: Bool = false

    /// Total report count from server.
    var totalCount: Int = 0

    /// Reports filtered by the selected status filter and optional type filter.
    var filteredReports: [ClientReportResponse] {
        var result = reports
        if selectedFilter != .all {
            result = result.filter { $0.status == selectedFilter.rawValue }
        }
        if let typeFilter = selectedTypeFilter {
            result = result.filter { $0.reportTypeId == typeFilter }
        }
        return result
    }

    /// Mobile-optimized, non-archived report types available for submission.
    var mobileReportTypes: [ClientReportTypeDefinition] {
        reportTypes.filter { $0.mobileOptimized && !$0.isArchived }
    }

    /// Whether typed report creation is available (at least one mobile-optimized type exists).
    var hasTypedReports: Bool {
        !mobileReportTypes.isEmpty
    }

    /// Resolve a report type label from its ID.
    func reportTypeLabel(for typeId: String?) -> String? {
        guard let typeId else { return nil }
        return reportTypes.first { $0.id == typeId }?.label
    }

    // MARK: - Initialization

    init(apiService: APIService, cryptoService: CryptoService, adminPubkeys: [String] = []) {
        self.apiService = apiService
        self.cryptoService = cryptoService
        self.adminPubkeys = adminPubkeys
    }

    // MARK: - Data Loading

    /// Load reports, categories, and report types from the API.
    /// Fetches CMS report types first (preferred), falling back to legacy endpoint.
    func loadReports() async {
        guard !isLoading else { return }
        isLoading = true
        errorMessage = nil

        async let reportsResult: Void = fetchReports()
        async let categoriesResult: Void = fetchCategories()
        async let typesResult: Void = fetchReportTypes()
        async let cmsTypesResult: Void = loadCmsReportTypes()

        await reportsResult
        await categoriesResult
        await typesResult
        await cmsTypesResult

        // Prefer CMS report types over legacy if available
        if !cmsReportTypes.isEmpty {
            reportTypes = cmsReportTypes
        }

        isLoading = false
    }

    /// Load CMS report type definitions from `GET /api/settings/cms/report-types`.
    /// This endpoint returns full definitions with CMS-specific fields.
    func loadCmsReportTypes() async {
        do {
            let types = try await apiService.fetchCmsReportTypes()
            cmsReportTypes = types
        } catch {
            // CMS endpoint may not be available — fall back to legacy types
            cmsReportTypes = []
        }
    }

    /// Refresh reports (pull-to-refresh).
    func refresh() async {
        isLoading = false
        await loadReports()
    }

    // MARK: - Report Creation (Legacy)

    /// Encrypt and create a new report (legacy category-based).
    ///
    /// - Parameters:
    ///   - title: The report title.
    ///   - category: Optional report category.
    ///   - body: The report body text.
    /// - Returns: `true` if creation succeeded.
    @discardableResult
    func createReport(title: String, category: String?, body: String) async -> Bool {
        isActionInProgress = true
        errorMessage = nil

        do {
            // Encrypt the body with E2EE envelope (same as notes)
            let encryptedNote = try cryptoService.encryptNote(payload: body, adminPubkeys: adminPubkeys)

            let request = CreateReportRequest(
                title: title,
                category: category,
                encryptedContent: encryptedNote.encryptedContent,
                authorEnvelope: NoteKeyEnvelope(
                    ephemeralPubkey: encryptedNote.authorEnvelope.ephemeralPubkey,
                    wrappedKey: encryptedNote.authorEnvelope.wrappedKey
                ),
                adminEnvelopes: encryptedNote.adminEnvelopes.map { env in
                    NoteRecipientEnvelope(
                        ephemeralPubkey: env.ephemeralPubkey,
                        pubkey: env.pubkey,
                        wrappedKey: env.wrappedKey
                    )
                }
            )

            let _: ClientReportResponse = try await apiService.request(
                method: "POST",
                path: apiService.hp("/api/reports"),
                body: request
            )

            // Haptic feedback on success
            let generator = UIImpactFeedbackGenerator(style: .medium)
            generator.impactOccurred()

            // Reload reports to include the new one
            await refresh()
            isActionInProgress = false
            return true
        } catch {
            errorMessage = error.localizedDescription
            isActionInProgress = false
            return false
        }
    }

    // MARK: - Typed Report Creation

    /// Encrypt and create a template-driven typed report.
    ///
    /// Field values are serialized as JSON, encrypted as a single E2EE payload,
    /// and sent with the `reportTypeId` in metadata.
    ///
    /// Uses `APIService.request(rawBody:)` to bypass the `convertToSnakeCase` encoder —
    /// the backend expects camelCase keys (`reportTypeId`, `encryptedContent`,
    /// `readerEnvelopes`).
    ///
    /// - Parameters:
    ///   - reportTypeId: The report type definition ID.
    ///   - title: The report title (derived from type label or first text field).
    ///   - fieldValues: Dictionary of field name to value, serialized as JSON for encryption.
    /// - Returns: `true` if creation succeeded.
    @discardableResult
    func createTypedReport(reportTypeId: String, title: String, fieldValues: [String: AnyCodableValue]) async -> Bool {
        isActionInProgress = true
        errorMessage = nil

        do {
            // Serialize field values as JSON for encryption
            let encoder = JSONEncoder()
            encoder.outputFormatting = .sortedKeys
            let fieldsData = try encoder.encode(fieldValues)
            let fieldsJSON = String(data: fieldsData, encoding: .utf8) ?? "{}"

            // Encrypt the fields JSON with per-message forward secrecy
            let encrypted = try cryptoService.encryptMessage(
                plaintext: fieldsJSON,
                readerPubkeys: adminPubkeys
            )

            // Encode body with a plain encoder (no snake_case conversion).
            // The backend expects camelCase keys (reportTypeId, encryptedContent,
            // readerEnvelopes), but APIService.encoder uses convertToSnakeCase.
            let body = CreateTypedReportRequest(
                title: title,
                category: nil,
                reportTypeId: reportTypeId,
                encryptedContent: encrypted.encryptedContent,
                readerEnvelopes: encrypted.envelopes
            )
            let plainEncoder = JSONEncoder()
            let rawBody = try plainEncoder.encode(body)

            let _: ClientReportResponse = try await apiService.request(
                method: "POST",
                path: apiService.hp("/api/reports"),
                rawBody: rawBody
            )

            // Haptic feedback on success
            let generator = UIImpactFeedbackGenerator(style: .medium)
            generator.impactOccurred()

            // Reload reports to include the new one
            await refresh()
            isActionInProgress = false
            return true
        } catch {
            errorMessage = error.localizedDescription
            isActionInProgress = false
            return false
        }
    }

    // MARK: - Report Actions

    /// Claim a waiting report by assigning it to the current user.
    func claimReport(id: String) async {
        guard let pubkey = cryptoService.pubkey else { return }
        isActionInProgress = true
        errorMessage = nil

        do {
            let request = ReportAssignRequest(assignTo: pubkey)
            let _: ClientReportResponse = try await apiService.request(
                method: "POST",
                path: apiService.hp("/api/reports/\(id)/assign"),
                body: request
            )

            let generator = UINotificationFeedbackGenerator()
            generator.notificationOccurred(.success)

            await refresh()
        } catch {
            errorMessage = error.localizedDescription
        }

        isActionInProgress = false
    }

    /// Close an active report.
    func closeReport(id: String) async {
        isActionInProgress = true
        errorMessage = nil

        do {
            let request = ReportUpdateRequest(status: "closed")
            let _: ClientReportResponse = try await apiService.request(
                method: "PATCH",
                path: apiService.hp("/api/reports/\(id)"),
                body: request
            )

            let generator = UINotificationFeedbackGenerator()
            generator.notificationOccurred(.success)

            await refresh()
        } catch {
            errorMessage = error.localizedDescription
        }

        isActionInProgress = false
    }

    // MARK: - Private Helpers

    private func fetchReports() async {
        do {
            let response: ReportsListResponse = try await apiService.request(
                method: "GET",
                path: apiService.hp("/api/reports") + "?limit=50"
            )
            reports = response.conversations
            totalCount = response.total
        } catch {
            if case APIError.noBaseURL = error {
                // Hub not configured — show empty state, no error
            } else {
                errorMessage = error.localizedDescription
            }
        }
    }

    private func fetchCategories() async {
        do {
            let response: ReportCategoriesResponse = try await apiService.request(
                method: "GET",
                path: apiService.hp("/api/reports/categories")
            )
            categories = response.categories
        } catch {
            // Categories are optional — silently continue without them
            categories = []
        }
    }

    private func fetchReportTypes() async {
        do {
            let response: ClientReportTypesResponse = try await apiService.request(
                method: "GET",
                path: apiService.hp("/api/reports/types")
            )
            reportTypes = response.reportTypes
        } catch {
            // Report types are optional — hub may not have any configured
            reportTypes = []
        }
    }
}
